### ISSUE TITLE
QOL: Add keybind for "Reset stuck cims and vehicles"

### ISSUE THREAD

**Issue by Karnor** [CitiesSkylinesMods/TMPE#1035]
I am having to use the Reset stuck cims/vehicles tool every two or three minutes. And the problem appears to be for cims, not vehicles, as I can find dormant zombie hordes all over the map, that starburst as soon as I use the tool, travelling in straight lines because where they're going, they don't need roads. After the zombie horde rushes off, the gameplay massively speeds up for a couple of minutes.

Presumably this has everything to do with the pf bug reports I have just read.

I used to follow one of the cims after they get unstuck to see if there are any obvious problems, but I gave up doing that when one of them literally crossed over the road to their destination, so like all of spitting distance away from his brain fart moment lol

**TL;DR** The point is that having to constantly go into the menu to use the tool every couple/few minutes is beyond a royal pita. It would be a massive QOL improvement to add this tool to a keybind. Please, please, please say you'll do it, please.

Question: As a dev why would I want to do this?
Answer: Because the required effort to QOL improvement ratio is a no brainer, and it would be trivial compared to the other open issues, so a mere coffee moment task :)

**Comment by Karnor** [comment]
BTW: I dunno if this helps you in tracking the pf problems, the same game save that grinds the windows client to a halt when the zombie hordes start congregating menacingly on some random segment/node, runs smoothly in the linux native client, albeit slow, but playable.

**Comment by krzychu124** [comment]
What could help is that savegame because that button was created as a last resort option if nothing else didn't help

**Comment by Karnor** [comment]
Oh sorry, I didn't think it would be needed for a feature request, but as I mentioned the behavioural difference between the linux and windows clients for a bug, I guess so, mb. I'll add it later today.

**Comment by Karnor** [comment]
I can't share the savegame on steam, I get a generic fail message when trying to share it, and I really cba to fight steam on yet another one of it's ambiguous errors.
I've shared the [savegame](https://drive.google.com/file/d/1so2g4y3ctGM2OJlSznPU2eQFS-K1E0i8/view?usp=sharing) on my google drive, hopefully that's good enough, or is there another way I can get the savegame to you?

**Comment by Karnor** [comment]
Interestingly my latest savegame was completely tanked after load, and the broken node tool found two broken PT lines. The first one seems to be where I randomly clicked the road while chasing down a moving bus with my mouse,  as it's the latest bus line, but the other one is named line zero, which while appropriate, is a non existent line in the bus lines list, and I'm not even sure you can have a legal line zero.

Having deleted both lines, which was a struggle itself in a frozen game, the game is now smooth again, and somewhat faster than it was before I raised this ticket, but still not as fast as I've seen it in other game sessions. It's slowing down again to the point I used the unstuck tool, but nowhere near as much as it was, so it seems line zero had a lot to do with things, but that line never showed up when I ran the tool before.

**Comment by krzychu124** [comment]
I've got the savegame, I'll take a look later today, thanks 😉

**Comment by Karnor** [comment]
Actually, I'm not sure if any changes related to this problem were in the latest update, but that map is running much better today. It's currently got a pop > 440K and running really smooth, with a very healthy simulation fps.

I still found a couple of zombie hordes on the map, and managed to get rid of them by adding a few walkways, as they always seem to congregate around junctions, and none of the junctions were restricted.

There was one particularly interesting horde, which was stuck trying to find a path to my industrial area, where they were all trying to go to work. A new walkway didn't fix the problem, so I put a new metro line into the area, and plopped a station right next to the horde, which they immediately used, so problem solved.
But the interesting point is that now I can see them rushing over the footbridge to where they were getting stuck, but they run straight past the metro station to the original stuck point, then when they get there, they do a u-turn and run down the metro station.
So it seems that even though they now have a path to their destination, they still have that original end of path point in their revised path. All the other hordes I fixed don't exhibit this behaviour, though that might be because the fix, i.e. walkway, etc, is closer to their original stuck point.
Dunno if that helps, or if you were aware of that sort of behaviour already, but I thought it was an interesting factor.

Also the strange thing is that there was already a point to point train line going into the area, and the starting station was literally right behind where they were congregating, but they just ignored it.

So it looks like my issue with the pf code is solved, and I am better versed in the problem to be able to spot and fix the issues.

But I still think the unstuck code should be assigned to a hotkey. Unsticking them, to see them run across the map to their destinations, and from other directions, helped me solve the problems, and as I said, having to repeatedly go in to the menu to use the tool is a pita, and can be quite tedious when you're trying to solve multiple problems.

Up to you if you close this ticket now as _no action required_ though.

**Comment by krzychu124** [comment]
Could you record short clip how it looks like? I tested your savegame yesterday and I didn't spot anything weird there 🤔 

I still think that better would be just fix the cause of stuck instead of repeatedly hit unstuck button/hotkey. 

Each citizen/vehicle is using one or multiple `PathUnit`(collection of waypoints) which may lead to weird issues like disconnecting public transport stops or not spawning vehicles if game hit the limit.

**Comment by krzychu124** [comment]
[Unrelated issue]
I noticed something weird with bikes:
Whenever cim wants to turn right(far turn) is despawned for no reason, also path looks wrong. It might be asset issue

![ezgif com-gif-maker (1)](https://user-images.githubusercontent.com/19638970/99889286-d9a04880-2c53-11eb-9c98-0084f275b8b0.gif)

**Comment by Karnor** [comment]
I get that you want to fix the bug, and yes it should be, but I'm not a dev trying to do that, I am simply trying to fix individual pf bugs in my games so I can play, which like I said would be massively helpful if there was a keybind for un-sticking the hordes. I assume this would be beneficial to everyone facing the same problems, but sure if it's just for my benefit then not worth the effort.

I haven't noticed the right turn problem, but I stopped using that network due to trams being in the middle of the road. So I swapped to one that has trams on the outside in another sandbox, but I noticed that for some inexplicable reason all the left turns were turned off, which caused stupid traffic routing problems. Not sure if that is related to the asset though.

grrr, I hate steam!!! I hosed that savegame, and cannot load the linked savegame back into my game. I put it in my local save folder, and in the local cloud folder, but steam refuses to sync, and CS doesn't see it in my local folder, so I can't run that game again :(

However, I have a savegame from later in that sandbox's evolution at 187K pop, which has the same problem, so [I'm sending you a link to a folder](https://drive.google.com/drive/folders/1UYNqEe1shvMjVcq8S0Qzyf646g0wiOci?usp=sharing) that has the savegame, and an accompanying video, plus the loading screen log for list of assets, and mod compatibility log for list of mods. It also has my player.log in case there's anything of use in there.

**NOTE:** The video size exceeds the limit that google drive will do a virus scan on before downloading. _I promise you the file is clean_, but you should do a scan your end anyway.

In the video I forgot to record the cursor, so my attempts at hovering over relevant points failed, but you will see the problem. I rotate the camera at first load so you can see where the location is, which is slightly right off the station if facing the clock tower.

- In the first horde I release, you should see one of them highlighted is going to the clock tower, which is just over the road and slightly left of her starting position. You should also notice one of them is going to work at the station which is literally just down the road to the left of his starting position.
- After I release the first horde, I go back to the same spot where you'll see they start massing again instantly.
- At seven minutes in the video I release the last horde, with a particular cim highlighted that works at the station. When I release the horde, I follow him, and of course he walks straight down the road to the station.
- If I leave the horde alone, it will keep on building up in numbers until the game starts to crawl. That horde by the station, is the one I mentioned in my previous comment, that I fixed by adding a metro line from where they mass, directly to the industrial area, which is already served by a point to point line from the station they are down the road from. No idea why that fixes the path for the station workers, or the cims visiting the clock tower though.

Anyway, hope that helps.

Now I have to try and work out why I can't load save games that I add to the save folder. How did you do it?

**Comment by krzychu124** [comment]
Sorry for late response, I was a bit busy.

Anyways, that's what I thought. 
It's the problem with the base game. Citizens stuck there because game restrict max distance between lanes people can walk. Each segment has lanes of length of the segment. Intersection area doesn't have lanes at all, every transition is calculated automatically. 

The problem is that game set transition length limit (64 units) in `CitizenAI` but not in PathFinder(PF), so PF can easily find path (it thinks it is correct, because it has no idea about that limit upon searching), but Citizen immediately invalidates it and requests new path, then, obviously PF will find "new path", but since Citizen didn't move and still wants to go to the same spot as previously, path returned by PF is exactly the same, so Citizen again invalidates it, and that process goes forever, as you may notice(or you hit reset stuck cim)... 
When citizen is moving using bike, `BikeAI` also invalidates path(the same code), but the difference is that instead of stopping and waiting for new path, bike is depawned, along with the driver (harder to spot).

I did few experiments, patched that limit in code and it looks like everything fixed itself. I did a bit more tests with some crazy configuration and I didn't spot anything weird. I'll add that code to TM:PE. Better would be a separate mod, but I don't have time to create it and I think that most of users are using TM:PE so it's not bad idea to have it there. 

Summarizing, it's vanilla game problem reveals when you use very wide networks, basically anything wider than 11 lanes (BadPeanut network you use has 12 lanes) can cause issue on outer lanes. Problem doesn't affect vehicles, because none of VehicleAI's check for transition length between lanes on intersections.

**Comment by Karnor** [comment]
Cool, so I learned something, i.e. lanes length, and you developed a potential solution. Thanks for the investigation/fix.
As Sod's law would have it, I have actually moved away from using the BadPeanut network because it was just overkill for anything I was doing lol
